File: src/lib/routing/outAndBackGenerator.ts
import type { UserPreferences } from "@/types/route";
import type { RouteGraph } from "./graph";
import { combinePaths, findShortestPath, reversePath } from "./graph";
import { findOutAndBackDestinations, getStartNode } from "./candidateGenerator";
import { resolveTargetDistanceKm } from "./geoUtils";
import { buildRouteCandidate } from "./routeAnalyzer";

export function generateOutAndBackRoutes(preferences: UserPreferences, graph: RouteGraph) {
  const startNode = getStartNode(graph, preferences);
  const targetDistanceM = resolveTargetDistanceKm(preferences) * 1000;
  const destinations = findOutAndBackDestinations(preferences, graph, startNode);

  const candidates = destinations
    .map((destination) => {
      const outbound = findShortestPath(graph, startNode.id, destination.id, preferences);
      if (!outbound) return null;
      const reverseError = Math.abs(outbound.distanceM * 2 - targetDistanceM);
      return { outbound, destination, reverseError };
    })
    .filter((c): c is NonNullable<typeof c> => c !== null);

  // Closest to target round-trip first.
  candidates.sort((a, b) => a.reverseError - b.reverseError);

  return candidates.map((candidate, index) => {
    const { outbound, destination } = candidate;
    // Out-and-back semantics: the return retraces the outbound exactly so
    // total distance is deterministic (2 × outbound) and the runner sees the
    // same scenery in reverse.
    const path = combinePaths(graph, [outbound, reversePath(outbound)]);

    return buildRouteCandidate(
      {
        id: `generated-out-back-${index + 1}`,
        name: "Out and Back",
        activity: preferences.activity,
        routeType: "out_and_back",
        path,
        waypoints: [destination.point],
        strategy: "recommended",
      },
      preferences,
    );
  });
}

File: services/routecraft-api/src/main/java/com/routecraft/api/routing/generator/OutAndBackGenerator.java
package com.routecraft.api.routing.generator;

import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.Optional;

import com.routecraft.api.routing.graph.GeoUtils;
import com.routecraft.api.routing.graph.PathResult;
import com.routecraft.api.routing.graph.Pathfinding;
import com.routecraft.api.routing.graph.RouteGraph;
import com.routecraft.api.routing.graph.RouteNode;
import com.routecraft.api.routing.model.RouteType;
import com.routecraft.api.routing.model.UserPreferences;

public final class OutAndBackGenerator {

    private OutAndBackGenerator() {
    }

    public static List<GeneratedRouteCandidate> generate(UserPreferences preferences, RouteGraph graph) {
        RouteNode startNode = CandidateSelector.getStartNode(graph, preferences);
        double targetDistanceM = GeoUtils.resolveTargetDistanceKm(preferences) * 1000;
        Pathfinding.ShortestPathTree searchTree =
                Pathfinding.buildShortestPathTree(graph, startNode.id(), preferences);
        List<RouteNode> destinations = CandidateSelector.findOutAndBackDestinations(
                preferences, graph, startNode, searchTree.pathDistancesM());

        record Outbound(PathResult outbound, RouteNode destination, double reverseError) {}
        List<Outbound> built = new ArrayList<>();
        for (RouteNode destination : destinations) {
            Optional<PathResult> outbound = searchTree.pathTo(graph, destination.id());
            if (outbound.isEmpty()) continue;
            double err = Math.abs(outbound.get().distanceM() * 2 - targetDistanceM);
            built.add(new Outbound(outbound.get(), destination, err));
        }
        built.sort(Comparator.comparingDouble(Outbound::reverseError));

        List<GeneratedRouteCandidate> result = new ArrayList<>();
        for (int i = 0; i < built.size(); i++) {
            Outbound entry = built.get(i);
            PathResult path = Pathfinding.combinePaths(graph, List.of(entry.outbound, Pathfinding.reversePath(entry.outbound)));
            RouteDraft draft = new RouteDraft(
                    "generated-out-back-" + (i + 1),
                    "Out and Back",
                    preferences.activity(),
                    RouteType.OUT_AND_BACK,
                    path,
                    List.of(entry.destination.point()),
                    RouteStrategy.RECOMMENDED,
                    null);
            result.add(RouteAnalyzer.buildRouteCandidate(draft, preferences));
        }
        return result;
    }
}
